stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -118,381 +48,389 @@ struct compute_dims<Eigen::Matrix<T, R, C>> {
 
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const std::vector<double>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  /**
+   * Return an eigen matrix of the adjoints
+   * @tparam RetType Not called by user. Default of the return type of
+   * operator() of the functor F allows for SFINAE to choose the correct
+   * specialization given the required output type.
+   */
+  template <typename RetType = FReturnType, require_eigen_t<RetType>* = nullptr>
+  inline auto y_adj() const {
+    Eigen::Matrix<double, FReturnType::RowsAtCompileTime,
+                  FReturnType::ColsAtCompileTime>
+        var_y(dims_[0], dims_[1]);
+    const size_t iter_size{dims_[0] * dims_[1]};
+    for (size_t i = 0; i < iter_size; ++i) {
+      var_y(i) = y_vi_[i]->adj_;
+    }
+    return var_y;
   }
-
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const std::vector<int>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  /**
+   * Copy the vari memory from the input argument
+   * @tparam EigMat A type inheriting from `EigenBase` with a scalar type of
+   * vari.
+   * @param x An eigen input argument
+   */
+  template <typename EigMat, require_eigen_vt<is_var, EigMat>* = nullptr>
+  inline auto prepare_x_vis_impl(EigMat&& x) const {
+    ref_type_t<EigMat> x_ref(std::forward<EigMat>(x));
+    vari** y
+        = ChainableStack::instance_->memalloc_.alloc_array<vari*>(x.size());
+    for (int i = 0; i < x_ref.size(); ++i) {
+      y[i] = x_ref.coeffRef(i).vi_;
+    }
+    return y;
   }
-
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const var& x, const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    count += 1;
-    return count_memory(count, args...);
+  /**
+   * Copy the vari memory from the input argument
+   * @param x An std vector of var types
+   */
+  inline auto prepare_x_vis_impl(const std::vector<var>& x) const {
+    vari** y
+        = ChainableStack::instance_->memalloc_.alloc_array<vari*>(x.size());
+    for (size_t i = 0; i < x.size(); ++i) {
+      y[i] = x[i].vi_;
+    }
+    return y;
   }
 
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const double& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  /**
+   * Copy the vari memory from the input argument
+   * @param x A var
+   */
+  inline auto prepare_x_vis_impl(const var& x) const {
+    vari* y = ChainableStack::instance_->memalloc_.alloc_array<vari>(1);
```

Comment:
Is the allocation necessary here? Can we just return x.vi_?

Reply:
Oh! I assumed we wanted a hard copy here. If not then yes we can just return x.vi_!

Reply:
Well I didn't fully think through this lol. I just remember removing this malloc in the other adj_jac_apply refactor.

Reply:
Ah okay so yes for this one with a `var` we can but the rest need to allocate

https://github.com/stan-dev/math/pull/1928/files#diff-0dc9dae1b5c3968c72caf717596dfb2fR165